src/conventions/h-group/interpret-play.js

```diff
@@ -69,6 +69,10 @@ export function interpret_play(game, action) {
 			if (game.rewind(card.drawn_index, { type: 'identify', order, playerIndex, suitIndex, rank }))
 				return;
 		}
+		// Possible self connections can be resolved.
+		// TODO: Should only remove them if the card may not be queued behind other plays.
```

Comment:
This, along with the similar line in `interpret-discard.js`, are a bit worrisome to me. I guess this might be a point in favour of the waiting connection method; all of the logic for determining whether a play/discard should cancel a possible connection is already there, and everyone else will be using that to determine whether it was an allowable defer.

I think the most common cases are when the "promised" card isn't playable yet (still requires others to play first) and when there's at least 1 connection possibility that isn't playable yet, but there are a few rarer ones too that we should be careful of.

Reply:
Yup, using a waiting connection might be better. In terms of how I'd start on this, what's the cleanest way of introducing this? The hacky way would be to mostly use the current logic for identifying a self_connection on the connection / focused card but add a connection instead of setting the self_connection property on the card. The less hacky way would be to actually pick up the connection and assign inferred card identities based on it, but the code currently only looks for self connections when other connections aren't found so I'm not quite sure what the cleanest way to do this is.

Reply:
My original idea was adding back the fake symmetric connections:
- Always look for symmetric connections in `resolve_clue()` except when we are the target, making sure to filter out save possibilities (around line 126 of interpret-clue.js, maybe pass in save possibilities to `find_symmetric_connections()`)
- Allow finding fake connections in `find_symmetric connections()` instead of skipping them, but maybe marking them with some flag (around line 133 of connection-helper.js)
- Modifying `thinksPlayables()`, etc. to check if there is a marked waiting connection involving the card
- Ensure other places using waiting connections only examine such marked connections if necessary

This wouldn't require modifying FocusPossibility or Connection objects at all, and I wouldn't write notes based on these; the bot already doesn't write notes on symmetric connections. It's almost completely disjoint from what you have now, though (and I found a few bugs while trying to briefly implement this, so might be quite a bit more work). :pensive: 

Something similar you could do (not sure if cleaner, sounds like the first approach you described) is to add a second focus possibility if a self-connection is possible, like how a second focus possibility is added if a finesse is possible. I don't think you should only add the connection to the existing focus possibility, since in cases where others react first the bot will "miss" such reactions. In `resolve_clue()` you'd probably have to identify the extra focus possibilities and treat them differently from "normal" ones, and you can create the waiting connection from there. At least, I think this could work, but it's kind of hard to think of all the cases without trying it.

I'm not sure I understand your second proposal, what do you mean by "pick up" the connection?

Reply:
> Always look for symmetric connections in resolve_clue() except when we are the target, making sure to filter out save possibilities (around line 126 of interpret-clue.js, maybe pass in save possibilities to find_symmetric_connections())
> Allow finding fake connections in find_symmetric connections() instead of skipping them, but maybe marking them with some flag (around line 133 of connection-helper.js)

The check in `own_finesses.js` https://github.com/WillFlame14/hanabi-bot/blob/master/src/conventions/h-group/clue-interpretation/own-finesses.js#L74 would need to be changed to I guess both look in our hand and other people's hands, but I don't think the code is currently expecting to find multiple possible cards per connecting card - i.e. we usually stop at the first. Maybe I'm missing something here. 

> I'm not sure I understand your second proposal, what do you mean by "pick up" the connection?

Right now in assign_connections I assign the self_connection boolean from the connection to the card, instead I would keep enough information in the self_connection property to add a waiting connection in assign_connections.

Reply:
I'll give your suggestion a try. I do think it's probably the cleanest if we generate and manage these connections the same way and will avoid a lot of the problems the current naive approach has.

Reply:
> The check in own_finesses.js https://github.com/WillFlame14/hanabi-bot/blob/master/src/conventions/h-group/clue-interpretation/own-finesses.js#L74 would need to be changed to I guess both look in our hand and other people's hands, but I don't think the code is currently expecting to find multiple possible cards per connecting card - i.e. we usually stop at the first. Maybe I'm missing something here.

Hmm, that's right. There's actually two new types of symmetric connections that are covered by your "self-connections": ones where we can fill in a missing finesse, and another where we have an ambiguous duplicate. The former can be covered by "fake" connections, but you correctly point out that we need to cover the latter as well.

A "normal" symmetric connection (currently already implemented):
- Only r1, y1 are played.
- Cathy clues 4 to Donald (touching r4).

Even though we know the finesse is red (2 blind plays from us), we know that Donald thinks that it could be yellow (2 blind plays from us).

A "fake" symmetric connection:
- r3 is played, possibly some others.
- Cathy clues 4 to Donald (touching r4).

Even though we know this is a direct play clue (0 blind plays from us), we should consider that the combination of some cards in our finesse position + others can form a connection that Donald might interpret (1+ blind plays from us).

An "ambiguous" symmetric connection:
- No cards are played.
- Bob clues green to Donald (touching g1).
- Cathy has g1 on finesse position and does something unrelated on her turn.

Even though we know this is a direct play clue (0 blind plays from us), Donald could think of connecting using Cathy's g1 -> g2, *and* Cathy already demonstrated that she didn't play, we could *still also* have g1 on finesse position that Donald might have to wait for. I think this case is very rare since it can only happen with 1s and everyone is aligned in order where Donald has already waited for Cathy but still needs to wait for us, but I guess it can happen. (This can't happen with ranks other than 1 because there are only 2 copies: if Donald is clued a 4 and Cathy has a 3, then Cathy deferring will cause a rewind so we know it's ambiguous on us. If Donald is clued a 3 and Cathy has a 3, then we cannot have the 3rd copy of the 3.)

So I guess yeah, we'd need to do something like modify `connect()` when finding own finesses to allow extra uses of our hand even when we can see a copy (and we could have such a card), or write a new `find_ambiguous_connections()` method. Note that `find_own_finesses()` is also called when we are the clue target and correctly identifies all types of possible connections, so I want to try avoiding unnecessary work if possible (maybe an if statement is enough?).

Reply:
On further thought, I think the "fake + ambiguous" situation can happen with ranks other than 1, since Cathy's action will not cause a rewind if the symmetric connection was fake. For example:
- r2, g3 is played.
- Bob clues 4 to Donald (touching g4).
- Cathy has r3 on finesse position and does something unrelated on her turn.

Then after Cathy doesn't play, we could still have a copy of r3 that Donald needs to wait for. So I guess we can only use the general check that our finesse position needs to have that identity as a possibility.

Reply:
> since Cathy's action will not cause a rewind if the symmetric connection was fake.

Should we rewind though? Couldn't this result in us finding the potential connection in our own hand after Cathy doesn't play?

Reply:
Right. If we rewind, then we would find the connection in our own hand afterward as a normal fake connection, since Cathy's finesse position would be ignored. I'm worried that this will cause a large amount of possibly-unnecessary rewinds though, since we would need to rewind for every fake connection to make sure we don't have an ambiguous card. I don't know, maybe I'm overestimating how common this case is in the first place.

Reply:
I've tried to implement the changes here though this is still failing a few tests so not ready to land yet. In particular it's not detecting the unsafe situation outlined earlier which is tested in `does not expect a play when it could be deferring playing into a finesse`. I haven't had time to continue digging into why exactly yet but figured I'd keep the branch up to date and push the code in case you have ideas or if i didn't implement your suggestions correctly.

Reply:
There were some logical issues with filtering symmetric connections that I fixed, it passes all tests now but doesn't include rewinding for fake ambiguous connections (might also have other bugs).

Reply:
Thanks so much!

> but doesn't include rewinding for fake ambiguous connections (might also have other bugs).

So I added a test for it (see latest commit), and it in fact does rewind and recognize the potential fake self connection.

Reply:
Ah right, that's because fake symmetric connections are currently treated almost the same as normal symmetric connections. There might be some hidden issues with this too that crop up later, I'm not really sure off the top of my head.

One thing I did notice is that even if Alice discards in the test, the fake symmetric connection isn't removed, due to being part of a possible [*Ambiguous Finesse* where another player may need to pass back to us](https://hanabi.github.io/extras/special-finesses/#the-ambiguous-finesse-pass-back). I think there are some ordering problems with this (i.e. restrict the if statement further by forcing players to be in a certain order), or maybe we could just pretend pass-backs can't occur with fake connections.

Reply:
> One thing I did notice is that even if Alice discards in the test, the fake symmetric connection isn't removed

I believe you there may still be an error case here, but ... the test says it is removed after alice discards:
https://github.com/WillFlame14/hanabi-bot/pull/224/commits/e52758fa2bd0d37b89faf918730edae6a1148ea1#diff-8b2dcfdd78aaa08351f1da52dc4392deabd2d124696b4b2b7fe82296f0944a1fR222

Reply:
I was looking at the debug logs and didn't examine what happened too carefully: it looks like the waiting connection is removed, but inferences aren't unwritten. Which is fine, since inferences aren't written/unwritten for symmetric connections in the first place. I'm not sure I completely remember the specifics of this case right now anyways, but we can look into it after something weird happens.